src/pudl/extract/eiaaeo.py

```diff
@@ -0,0 +1,252 @@
+                "case": case_name,
+            }
+            for d in series["data"]
+        )
+        return records
+
+    def get_table(self, table_number: int) -> pd.DataFrame:
+        """Get a specific table number as a DataFrame."""
+        matching_category_ids = {
+            n_id
+            for n_id in self.graph
+            if self.__sanitize(self.graph.nodes.get(n_id).get("name", "")).startswith(
+                f"table_{table_number}_"
+            )
+        }
+
+        # many series belong to more than one category, hence turning this into a set
+        matching_series = set(
+            itertools.chain.from_iterable(
+                self.graph.nodes[n_id].get("childseries", [])
+                for n_id in matching_category_ids
+            )
+        )
+
+        series_records = itertools.chain.from_iterable(
+            self.__series_to_records(series_id, potential_parents=matching_category_ids)
+            for series_id in matching_series
+        )
+        return AEOTable(pd.DataFrame.from_records(series_records))
+
+
+@multi_asset(
+    outs={
+        "raw_eiaaeo__electric_power_projections_regional": AssetOut(is_required=False),
+        "raw_eiaaeo__renewable_energy_regional": AssetOut(is_required=False),
+    },
+    can_subset=True,
+    required_resource_keys={"datastore"},
+)
+def raw_eiaaeo(context: AssetExecutionContext):
+    """Extract tables from EIA's Annual Energy Outlook.
+
+    We first extract a taxonomy from the AEO JSON blob, which connects
+    individual data series to "categories". Some categories are associated with
+    a specific table; others are associated with an AEO case or subject.
+
+    The AEO cases are different scenarios such as "High Economic Growth" or
+    "High Oil Price." They include "Reference" and "2022 AEO reference case" as
+    well.
+
+    The AEO subjects are only used for filtering which tables are relevant to
+    which subjects, e.g. "Table 54 is relevant to Energy Prices." So we ignore
+    those right now.
+
+    The series each have their own timeseries data, as well as some metadata
+    such as a series name and units. Many different dimensions can be inferred
+    from the series names, but the data is somewhat heterogeneous so we do not
+    try to infer those here and leave that to the transformation step.
+    """
+    name_to_number = {
+        "raw_eiaaeo__electric_power_projections_regional": 54,
+        "raw_eiaaeo__renewable_energy_regional": 56,
+    }
+    ds = context.resources.datastore
+    year = 2023
+    filename = f"AEO{year}.txt"
+
+    with ds.get_zipfile_resource("eiaaeo", year=year).open(
+        filename, mode="r"
+    ) as aeo_raw:
+        taxonomy = AEOTaxonomy(io.TextIOWrapper(aeo_raw))
+
+    selected = context.op_execution_context.selected_output_names
+    for asset_name in selected:
+        yield Output(
+            value=taxonomy.get_table(name_to_number[asset_name]), output_name=asset_name
+        )
```

Comment:
Is this generator construct to enable subsetting of the multi-asset?

Reply:
Yep!